### PR TITLE
Connect manually created cases to an order

### PR DESCRIPTION
**Pull request by islean**
## Description

Cases created via the CLI do not get an entry in the `order_case` table, which blocks analyses from being uploaded. This PR connects the case to the order specified via the ticket_id.

### Fixed

- Cases added via the CLI gets connected to the order related to the ticket_id.

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b fix-add-case -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/cli/add.py [hunk 1]
```diff
@@ -278,6 +279,8 @@ def add_case(
         priority=priority,
         ticket=ticket,
     )
+    order: Order = status_db.get_order_by_ticket_id(int(ticket))
+    new_case.orders.append(order)
```
Comment (islean) [inline cg/cli/add.py:283]: Do we want support for creating an order if it is not existing already? Or is it better to get an error if that happens?

Comment (seallard) [reply cg/cli/add.py:283]: I think it is better if an error is thrown. Seems the most intuitive to me if orders should only are created when a customer places an order via the order portal

**Review by seallard — APPROVED** [review]
LGTM! Might need some hacking in the test suite to get them to pass

Side track: does it make sense to allow adding cases to an order after it has been placed?

**Comment by islean** [comment]
> LGTM! Might need some hacking in the test suite to get them to pass
> 
> Side track: does it make sense to allow adding cases to an order after it has been placed? Maybe these cases are just for dev purposes

Probably not but obviously there are use cases for it currently. Maybe something to refine for the future?

File: cg/cli/add.py (hunk 1 above)
Comment (ChrOertlin) [reply cg/cli/add.py:283]: I don't exactly know the extent of the new order flow to be honest. If a new order is created does that trigger something else somewhere, for example invoicing etc? Or are we planning to do so?

How is this used mostly? I guess in the context of the bug it was likely some development thing, should this be considered?

File: cg/cli/add.py [hunk 2]
```diff
@@ -278,6 +279,8 @@ def add_case(
         priority=priority,
         ticket=ticket,
     )
+    order: Order = status_db.get_order_by_ticket_id(int(ticket))
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/cli/add.py:282]: I think we should take a look at this though. We might be representing the same thing in multiple tables as different types now.

File: cg/cli/add.py (hunk 1 above)
Comment (islean) [reply cg/cli/add.py:283]: The Order table was introduced to track analyses in trailblazer which belonged to the same order/ticket. Given that we soon will switch ticketing system and that cases can belong to many tickets either way, it seemed reasonable to have a many-to-many solution. 

There are quite a lot of areas where we could use the Order entries, but as of yet, it is only used for new views in CiGRID/Trailblazer. I would like to have a project in the future where we consider other areas which would benefit from using the table though.

In short, a new order does not trigger anything, but hopefully we could find some use cases in the future.

Comment (islean) [reply cg/cli/add.py:283]: Regarding test cases, it is a good point. Currently there is no available option to create an order via the CLI, so the tester would need to either connect to the database or submit an orderform in stage. Maybe that speaks for the option that a missing order should be created? Maybe with a click.confirm and warning message?

File: cg/cli/add.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/cli/add.py:282]: I recored this here as an issue https://github.com/Clinical-Genomics/cg/issues/3024

File: cg/cli/add.py (hunk 1 above)
Comment (ChrOertlin) [reply cg/cli/add.py:283]: Yes, I think so. 

> Regarding test cases, it is a good point. Currently there is no available option to create an order via the CLI, so the tester would need to either connect to the database or submit an orderform in stage. Maybe that speaks for the option that a missing order should be created? Maybe with a click.confirm and warning message?

Yes, I think so. But if orders will trigger something in future this would require all users to be aware of this and knowledge on other parts of the system. We likely need to document this somehow then.

Comment (islean) [reply cg/cli/add.py:283]: I guess? I mean it is slightly weird already that you can set a non-existent ticket on your case. Currently it is just a database model intended to contain data related to an order/ticket. Oh well, for a future discussion!

**Comment by islean** [comment]
Tested adding to existing order and order_case entry was created:

image
image

Tested adding to non-existent order and got the option to proceed. Aborted if no confirmation was given:

```
θ65° 8d [hiseq.clinical@hasta:~] [S_base] 22s $ cg add case -a mip-dna -dd analysis -t 167890 cust000 justguppy
No order found with ticket_id 167890
No order found for the given ticket, proceed? [y/N]: N
Aborted!
θ67° 8d [hiseq.clinical@hasta:~] [S_base] 31s 1 $ 
```

When confirmed, it works and an order was created tied to the case:

```
θ68° 8d [hiseq.clinical@hasta:~] [S_base] 37s 1 $ cg add case -a mip-dna -dd analysis -t 167890 cust000 just-guppy
No order found with ticket_id 167890
No order found for the given ticket, proceed? [y/N]: y
polishedeagle: new case added
θ66° 8d [hiseq.clinical@hasta:~] [S_base] 25s $ 
```

**Comment by islean** [comment]
Deployed to production:

```
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/hiseq.clinical
done.
Log deploy... done.
cg, version 59.10.8
θ65° 7d [hiseq.clinical@hasta:~] [P_base] 2m30s $
```
